AlpsSDK/ProximityHandler.swift

```diff
@@ -10,224 +10,127 @@ import Foundation
 import CoreLocation
 import Alps
 
-class ProximityHandler {
-    
-    var contextManager: ContextManager!
-    static var immediateTrigger: [String: ProximityEvent] = [:]
-    static var nearTrigger: [String: ProximityEvent] = [:]
-    static var farTrigger: [String: ProximityEvent] = [:]
-    static var unknownTrigger: [String: ProximityEvent] = [:]
-    var immediateBeacons: [String] = []
-    var nearBeacons: [String] = []
-    var farBeacons: [String] = []
-    var unknownBeacons: [String] = []
+protocol ProximityHandlerDelegate : class {
+    func didRangeBeacons(manager: ContextManager, beacons: [CLBeacon], knownBeacons: [IBeaconTriple])
+}
+
+class ProximityHandler: NSObject, ProximityHandlerDelegate {
     var refreshTimer: Int = 60 * 1000 // timer is in milliseconds
+    var beaconsDetected: [String: [IBeaconTriple]] = [:]
+    var beaconsTriggered: [String: [String: ProximityEvent]] = [:]
     
-    init(contextManager: ContextManager) {
-        self.contextManager = contextManager
+    override init() {
+        super.init()
+        beaconsTriggered["unknown"] = [String: ProximityEvent]()
+        beaconsTriggered["immediate"] = [String: ProximityEvent]()
+        beaconsTriggered["far"] = [String: ProximityEvent]()
+        beaconsTriggered["near"] = [String: ProximityEvent]()
     }
     
-    // Parsing beacons proximity relatively to the device that detect them
-    func parseBeaconsByProximity(_ beacons: [CLBeacon]) {
-        var ourBeacon: IBeaconTriple?
-        for beacon in beacons {
-            let b = syncBeacon(beacon: beacon)
-            if !b.isEmpty {
-                ourBeacon = b[0]
-            }
-            guard let deviceId = ourBeacon?.deviceId else {
-                return
-            }
-            removeDuplicate(clBeacon: beacon, deviceId: deviceId)
+    func didRangeBeacons(manager: ContextManager, beacons: [CLBeacon], knownBeacons: [IBeaconTriple]) {
+        let unknownBeacons = beacons.filter {$0.proximity == CLProximity.unknown}
+        let immediateBeacons = beacons.filter {$0.proximity == CLProximity.immediate}
+        let nearBeacons = beacons.filter {$0.proximity == CLProximity.near}
+        let farBeacons = beacons.filter {$0.proximity == CLProximity.far}
+        
+        // Filter with existing beacons in application
+        beaconsDetected["unknown"] = synchronizeBeacons(beacons: unknownBeacons, knownBeacons: knownBeacons)
+        beaconsDetected["immediate"] = synchronizeBeacons(beacons: immediateBeacons, knownBeacons: knownBeacons)
+        beaconsDetected["near"] = synchronizeBeacons(beacons: nearBeacons, knownBeacons: knownBeacons)
+        beaconsDetected["far"] = synchronizeBeacons(beacons: farBeacons, knownBeacons: knownBeacons)
+        
```

Comment:
Stage 2:
Synchronising with knownBeacons (all the beacons registered in the portal)

Note : means -> ContextManager.knownBeacons = BeaconRepositories.items
Correct me if wrong.

Reply:
That's sounds good.

We should think about getting beacons. Who and how often will refresh them?

Reply:
What do you mean with who getting beacons ?

You mean that the developer could access the beaconsDetected[CLProximity] to do something with it ?

Theses arrays are refreshed every time didRangeBeacons is called by locationManager. Almost every 1 second I would say.